stubs/icalendar/icalendar/tools.pyi

```diff
@@ -1,12 +1,23 @@
+import datetime
 from typing import Final
+from typing_extensions import TypeIs
+
+from pytz.tzinfo import BaseTzInfo
 
 from .prop import vText
 
-__all__ = ["UIDGenerator"]
+__all__ = ["UIDGenerator", "is_date", "is_datetime", "to_datetime", "is_pytz", "is_pytz_dt", "normalize_pytz"]
 
 class UIDGenerator:
     chars: Final[list[str]]
     @staticmethod
     def rnd_string(length: int = 16) -> str: ...
     @staticmethod
     def uid(host_name: str = "example.com", unique: str = "") -> vText: ...
+
+def is_date(dt: datetime.date) -> TypeIs[datetime.date]: ...  # and not datetime.date
+def is_datetime(dt: datetime.date) -> TypeIs[datetime.datetime]: ...
+def to_datetime(dt: datetime.date) -> datetime.datetime: ...
+def is_pytz(tz: datetime.tzinfo) -> TypeIs[BaseTzInfo]: ...
+def is_pytz_dt(dt: datetime.date) -> TypeIs[datetime.datetime]: ...  # and dt.tzinfo is BaseTZInfo
```

Comment:
Should probably use TypeGuard here because of the extra condition.

Reply:
Why would `TypeGuard` be more appropriate? From a type checking perspective, this function is equivalent to `is_datetime` (since we can't really make any assertions about its fields).

Reply:
TypeIs[T] must always return True if and only if the argument is of type T; this is what allows narrowing in the case where it returns False. TypeGuard does not have this restriction.

Reply:
But that's the case here. `is_pytz_dt` is implemented as follows:

```python
def is_datetime(dt: date) -> bool:
    """Whether this is a date and not a datetime."""
    return isinstance(dt, datetime)

# ...

def is_pytz_dt(dt: date):
    """Whether the time requires localize() and normalize()."""
    return is_datetime(dt) and is_pytz(dt.tzinfo)
```

Reply:
So it doesn't return True for all datetimes, right?

The problem would be code like this:
```
class MyDate(date): pass # unrelated subclass of date

x: MyDate | datetime
if is_pytz_dt(x):
    reveal_type(x) # datetime, correct
else:
    reveal_type(x) # type checker will narrow to MyDate, but it might be a datetime with a non-pytz tzinfo
```

Reply:
I see, makes sense!